File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        List<List<Integer>> graph = new ArrayList<>();

        // 입력
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int v = Integer.parseInt(st.nextToken());

        for (int i = 0; i <= n; i++){
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i< m; i++){
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph.get(a).add(b);
            graph.get(b).add(a);
        }

        for (int i = 1; i <= n; i++){
            Collections.sort(graph.get(i));
        }

        boolean[] visited = new boolean[n+1];
        dfs(graph, visited, v);
        System.out.println();
        visited = new boolean[n+1];
        bfs(graph, visited, v);
    }

    static void dfs(List<List<Integer>> graph, boolean[] visited, int v){
        visited[v] = true;
        System.out.print(v + " ");
        for (int i : graph.get(v)){
            if (!visited[i]){
                dfs(graph, visited, i);
            }
        }
    }

    static void bfs(List<List<Integer>> graph, boolean[] visited, int v){
        Queue<Integer> q = new LinkedList<>();
        q.add(v);
        visited[v] = true;

        while (!q.isEmpty()){
            int cur = q.poll();
            System.out.print(cur + " ");
            for (int i : graph.get(cur)){
                if (!visited[i]){
                    q.add(i);
                    visited[i] = true;
                }
            }
        }
    }
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque
import sys
input = sys.stdin.readline

N, M, V = map(int, input().split())

# 맵 입력
graph = [[] for _ in range(N+1)]
for _ in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

# 방문 순서가 정점 번호가 작은 순서여야 하므로 정렬
for adj in graph:
    adj.sort()

# DFS
visited_dfs = [False] * (N+1)
dfs_result = []

def dfs(v):
    visited_dfs[v] = True
    dfs_result.append(v)
    for nv in graph[v]:
        if not visited_dfs[nv]:
            dfs(nv)

# BFS
visited_bfs = [False] * (N+1)
bfs_result = []

def bfs(start):
    queue = deque([start])
    visited_bfs[start] = True
    while queue:
        v = queue.popleft()
        bfs_result.append(v)
        for nv in graph[v]:
            if not visited_bfs[nv]:
                visited_bfs[nv] = True
                queue.append(nv)
dfs(V)
bfs(V)
print(*dfs_result)
print(*bfs_result)
